jetty-util/src/main/java/org/eclipse/jetty/util/PidFile.java

```diff
@@ -0,0 +1,69 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import org.eclipse.jetty.util.annotation.Name;
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.eclipse.jetty.util.thread.ShutdownThread;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.nio.file.StandardOpenOption.CREATE;
+import static java.nio.file.StandardOpenOption.TRUNCATE_EXISTING;
+import static java.nio.file.StandardOpenOption.WRITE;
+
+/**
+ * Create a PID file for the running process.
+ */
+public class PidFile extends AbstractLifeCycle
+{
+    private static final Logger LOG = LoggerFactory.getLogger(PidFile.class);
+
+    private final Path pidFile;
+
+    public PidFile(@Name("file") String filename)
+    {
+        pidFile = Paths.get(filename);
+        try
+        {
+            // Create the PID file as soon as possible.
+            // We don't want for doStart() as we want the PID creation to occur quickly for jetty.sh
+            long pid = ProcessHandle.current().pid();
+            Files.writeString(pidFile, Long.toString(pid), UTF_8, CREATE, WRITE, TRUNCATE_EXISTING);
+            ShutdownThread.register(this);
```

Comment:
You're registering a new `LifeCycle` instance in the shutdown thread every time an instance of this class is created without ever trying to clean that up or document it in any way.

This means the following valid-looking loop would eventually OOME:

```java
while (true)
{
 Server server = new Server(8080);
 server.addManaged(new PidFile()); // this writes the PID file
 server.start();
 server.stop(); // this deletes the PID file
}
```

At the very least, we should document that this class must not be used with embedded servers.

Reply:
Yeah, but that's not what PidFile does, it registers itself with the ShutdownThread, it's not a Server managed lifecycle.

Reply:
Can we have a static to make sure only 1 is every installed

Reply:
I don't understand these comments.
I don't see why we need this complexity.
Few other components have this new requirement.
There's also nothing special about this class that prevents an embedded server from using it (dropwizard, spark, spring-boot, runner should all be able to use it).

The PidFile is a ShutdownThread component, let me at least add javadoc to that point.  Hopefully that will address this requested change.

Reply:
@joakime we have few other components that register JVM hooks that are not deleted.
The problem with my suggestion is that a simple static boolean would not handle the plausible (if rare) case where two `PidFile` instances are created for different file names.

I think it is really rare that any stop/start cycle will create enough of these objects to ever be a problem... however, the complexity to have a static concurrent set of filenames for which a shutdown hook has already been registered is not that much.   If a filename is already in that static set, then simply do not register the shutdownThread.      Then you have covered the 0.0000001% case as well :)